src/core/service.c

```diff
@@ -1250,23 +1271,98 @@ static int service_coldplug(Unit *u) {
         return 0;
 }
 
+static int connect_unix_probe(Service *s, const OpenFile *of, int ofd) {
+        union sockaddr_union addr = {
+                .un.sun_family = AF_UNIX,
+        };
+        socklen_t sa_len;
+        int socket_types[] = { SOCK_DGRAM, SOCK_STREAM, SOCK_SEQPACKET };
+        int r;
+
+        assert(s);
+        assert(of);
+        assert(ofd >= 0);
+
+        r = sockaddr_un_set_path(&addr.un, FORMAT_PROC_FD_PATH(ofd));
+        if (r < 0)
+                return log_unit_error_errno(UNIT(s), r, "Failed to set sockaddr for %s: %m", of->path);
+
+        sa_len = r;
+
+        for (size_t i = 0; i < ELEMENTSOF(socket_types); i++) {
+                _cleanup_close_ int fd = -1;
+
+                fd = socket(AF_UNIX, socket_types[i] | SOCK_CLOEXEC, 0);
+                if (fd < 0)
+                        return log_unit_error_errno(
+                                        UNIT(s), errno, "Failed to create socket for %s: %m", of->path);
+
+                r = RET_NERRNO(connect(fd, &addr.sa, sa_len));
+                if (r >= 0)
+                        return TAKE_FD(fd);
+
+                if (r != -EPROTOTYPE)
```

Comment:
hmm, sure this is the error we see here? isn't that EINVAL?

Reply:
No i'ts definitely EPROTOTYPE :)